### ISSUE TITLE
[FEA] Refactor `strings::make_strings_children` and `strings:: make_strings_children_with_null_mask`

### ISSUE THREAD

**Issue by ttnghia** [rapidsai/cudf#8059]
Currently, `strings::make_strings_children` and `strings:: make_strings_children_with_null_mask` are located in the `src` folder (`src/strings/utilities.cuh`). However, since they are so convenient, they are gradually used outside of the `strings` namespace in many other places (such as in the `lists` namespace). 

As such, there is a need to move them into the `include/strings/detail/utilities.cuh`. **This is the first proposed item of this issue.**

I recently (https://github.com/rapidsai/cudf/pull/7929) added the second internal API (`strings:: make_strings_children_with_null_mask`) that compensates for the lacking ability to generate null mask and null count of the first API (`strings::make_strings_children`). In addition, the second API also allows executing the template functor at a different range from the output range (number of generating strings). This is important and very useful for the cases when we want to execute the functor in a transform-reduce style, instead of just transform.

Thus, it would be necessary to add an additional `exec_size` parameter to the API `strings::make_strings_children`, allowing to execute the template functor at a different range from the output range.  **This is the second proposed item of this issue.**

FYI, currently I have two PRs that depend on the changes proposed here:
 * https://github.com/rapidsai/cudf/pull/8046/files#diff-d5148f2f125b2e073936f356808e2700593126476ceceb92e275477e30c6fec5R178-R182
 * https://github.com/rapidsai/cudf/pull/8049/files#diff-e419dfe9cb79cb8941ffc72e759d4260f7e1ac22b08015838d603ff4a26806e9R259-R263

In order to reduce breaking when adding an additional parameter `exec_size` to `strings::make_strings_children`, I would suggest doing the following:
1. Change the interface of the existing function `strings::make_strings_children` to have an additional `exec_size` parameter:
```
make_strings_children(
  SizeAndExecuteFunction size_and_exec_fn,
  size_type exec_size,
  size_type strings_count,
  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
```
2. Add a new overloaded function that calls the function above:
```
make_strings_children(
  SizeAndExecuteFunction size_and_exec_fn,
  size_type strings_count,
  rmm::cuda_stream_view stream        = rmm::cuda_stream_default,
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource()) {
    return make_strings_children(size_and_exec_fn, strings_count, strings_count, stream, mr);
}
```

We also need to do the same for the API `make_strings_children_with_null_mask` so it can have two overloads and can be called differently, depending on the use case.

===========================================================================================
**In summary:**
1. Move `strings::make_strings_children` and `strings:: make_strings_children_with_null_mask` into the `include/strings/detail/utilities.cuh`.
2. Each of the two APIs should have two interfaces to allows executing the template functor in both transform and transform-reduce styles.

**Comment by ttnghia** [comment]
@davidwendt I created this issue so we can keep track of the matter. Please let me know if you have any opinion on this.
